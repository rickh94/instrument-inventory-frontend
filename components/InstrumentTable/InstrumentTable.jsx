import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { TableCell, Typography } from '@material-ui/core'
import {withStyles} from '@material-ui/styles'
import clsx from 'clsx'
import { AutoSizer, Column, Table } from 'react-virtualized'
import DownIcon from '@material-ui/icons/ArrowDropDown'
import UpIcon from '@material-ui/icons/ArrowDropUp'
import { stars, yesOrNo, truncateText } from '../../libs/display'
import { withRouter } from 'react-router-dom'

const styles = theme => ({
  flexContainer: {
    display: 'flex',
    alignItems: 'center',
    boxSizing: 'border-box',
  },
  tableRow: {
    cursor: 'pointer',
  },
  tableRowHover: {
    '&:hover': {
      backgroundColor: theme.palette.grey[200],
    },
  },
  tableCell: {
    flex: 1,
  },
  noClick: {
    cursor: 'initial',
  },
  icon: {
    position: 'relative',
    top: 2,
    // width: theme.typography.subtitle1.fontSize,
    // height: theme.typography.subtitle1.fontSize,
    width: 18,
    height: 18,
  },
})

class MuiVirtualizedTable extends React.PureComponent {
  static defaultProps = {
    headerHeight: 48,
    rowHeight: 48,
  }

  getRowClassName = ({ index }) => {
    const { classes, onRowClick } = this.props
    return clsx(classes.tableRow, classes.flexContainer, {
      [classes.tableRowHover]: index !== -1 && onRowClick != null,
    })
  }

  cellRenderer = ({ cellData, columnIndex, dataKey, ...other }) => {
    const { columns, classes, rowHeight, onRowClick } = this.props
    if (dataKey === 'condition' || dataKey === 'quality') {
      cellData = stars(cellData)
    } else if (
      dataKey === 'ready' ||
      dataKey === 'rosin' ||
      dataKey === 'bow' ||
      dataKey === 'shoulderRestEndpinRest' ||
      dataKey === 'gifted'
    ) {
      cellData = yesOrNo(cellData)
    } else if (dataKey === 'maintenanceNotes' || dataKey === 'conditionNotes') {
      cellData = truncateText(cellData, 15)
    } else if (dataKey === 'history' && cellData) {
      cellData = truncateText(cellData.join(', '), 30)
    }
    return (
      <TableCell
        component="div"
        className={clsx(classes.tableCell, classes.flexContainer, {
          [classes.noClick]: onRowClick == null,
        })}
        variant="body"
        style={{ height: rowHeight }}
        align={
          (columnIndex != null && columns[columnIndex].numberic) || false
            ? 'right'
            : 'left'
        }
      >
        {cellData}
      </TableCell>
    )
  }

  headerRenderer = ({ label, columnIndex, dataKey }) => {
    const { headerHeight, columns, classes, sortByField } = this.props
    let arrowIcon = ''
    if (this.props.sortBy === dataKey) {
      arrowIcon =
        this.props.sortDirection === 'DESC' ? (
          <DownIcon className={classes.icon} />
        ) : (
          <UpIcon className={classes.icon} />
        )
    }

    return (
      <TableCell
        component="div"
        className={clsx(classes.tableCell, classes.flexContainer, {
          [classes.noClick]: dataKey === 'history',
        })}
        variant="head"
        style={{ height: headerHeight }}
        align={columns[columnIndex].numeric || false ? 'right' : 'left'}
        onClick={sortByField(dataKey)}
      >
        <Typography variant="body2">
          {label} {arrowIcon}
        </Typography>
      </TableCell>
    )
  }

  render() {
    const { classes, columns, ...tableProps } = this.props

    return (
      <AutoSizer>
        {({ height, width }) => (
          <Table
            height={height}
            width={width}
            {...tableProps}
            rowClassName={this.getRowClassName}
          >
            {columns.map(({ dataKey, ...other }, index) => {
              return (
                <Column
                  key={dataKey}
                  headerRenderer={headerProps =>
                    this.headerRenderer({ ...headerProps, columnIndex: index })
                  }
                  className={classes.flexContainer}
                  cellRenderer={this.cellRenderer}
                  dataKey={dataKey}
                  {...other}
                />
              )
            })}
          </Table>
        )}
      </AutoSizer>
    )
  }
}

MuiVirtualizedTable.propTypes = {
  classes: PropTypes.object.isRequired,
  columns: PropTypes.arrayOf(PropTypes.object).isRequired,
  headerHeight: PropTypes.number,
  onRowClick: PropTypes.func,
  rowHeight: PropTypes.number,
  sortByField: PropTypes.func.isRequired,
}

const VirtualizedTable = withStyles(styles)(MuiVirtualizedTable)

const defaultForField = {
  number: '',
  type: '',
  size: '',
  assignedTo: '',
  location: '',
  condition: 0,
  quality: 0,
  gifted: false,
}

const compareField = (field, a, b, sortDirection) => {
  const aValue = a[field] || defaultForField[field]
  const bValue = b[field] || defaultForField[field]
  if (aValue < bValue) {
    return -sortDirection
  }
  if (bValue < aValue) {
    return sortDirection
  }
  return 0
}

export const InstrumentTable = ({ records, history }) => {
  const [rows, setRows] = useState(records)
  const [sortedBy, setSortedBy] = useState('')
  const [sortDirection, setSortDirection] = useState(1)

  const sortByField = field => () => {
    if (field === 'history') {
      return
    }
    let direction = sortDirection
    if (sortedBy === field) {
      setSortDirection(sortDirection => -sortDirection)
      direction = -direction
    } else {
      setSortDirection(1)
      direction = 1
    }
    setSortedBy(field)
    setRows(rows => {
      const newRows = rows.slice(0)
      newRows.sort((a, b) => compareField(field, a, b, direction))
      return newRows
    })
  }

  const onRowClick = ({ rowData }) => history.push(`/instrument/${rowData.id}`)

  return (
    <React.Fragment>
      <VirtualizedTable
        sortBy={sortedBy}
        sortDirection={sortDirection === 1 ? 'DESC' : 'ASC'}
        sortByField={sortByField}
        rowCount={rows.length}
        rowGetter={({ index }) => ({ ...rows[index] })}
        onRowClick={onRowClick}
        columns={[
          { width: 120, label: 'Number', dataKey: 'number' },
          { width: 140, label: 'Type', dataKey: 'type' },
          { width: 110, label: 'Size', dataKey: 'size' },
          { width: 180, label: 'Assigned To', dataKey: 'assignedTo' },
          { width: 210, label: 'Location', dataKey: 'location' },
          { width: 120, label: 'Condition', dataKey: 'condition' },
          { width: 110, label: 'Quality', dataKey: 'quality' },
          { width: 100, label: 'Gifted', dataKey: 'gifted' },
          { width: 200, label: 'History', dataKey: 'history' },
        ]}
        headerHeight={48}
        rowHeight={48}
      />
    </React.Fragment>
  )
}

InstrumentTable.propTypes = {
  history: PropTypes.object.isRequired,
  records: PropTypes.array.isRequired,
}

export default withRouter(InstrumentTable)
