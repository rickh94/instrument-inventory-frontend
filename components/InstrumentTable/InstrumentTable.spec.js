import React from 'react'
import { render } from '@testing-library/react'
import { InstrumentTable } from './InstrumentTable'

describe('<InstrumentTable />', () => {
  test('renders', () => {
    const { container } = render(
      <InstrumentTable records={[{}]} history={{ push: jest.fn() }} />
    )
    expect(container).toMatchSnapshot()
  })
})
