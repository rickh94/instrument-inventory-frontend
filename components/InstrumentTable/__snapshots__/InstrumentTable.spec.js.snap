// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InstrumentTable /> renders 1`] = `
<div>
  <div
    style="overflow: visible; height: 0px; width: 0px;"
  >
    <div
      aria-colcount="9"
      aria-rowcount="1"
      class="ReactVirtualized__Table"
      role="grid"
    >
      <div
        class="ReactVirtualized__Table__headerRow MuiVirtualizedTable-tableRow-2 MuiVirtualizedTable-flexContainer-1"
        role="row"
        style="height: 48px; overflow: hidden; padding-right: 0px; width: 0px;"
      >
        <div
          class="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style="flex: 0 1 120px;"
        >
          <div
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiVirtualizedTable-tableCell-4 MuiVirtualizedTable-flexContainer-1"
            style="height: 48px;"
          >
            <p
              class="MuiTypography-root MuiTypography-body2"
            >
              Number
               
              
            </p>
          </div>
        </div>
        <div
          class="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style="flex: 0 1 140px;"
        >
          <div
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiVirtualizedTable-tableCell-4 MuiVirtualizedTable-flexContainer-1"
            style="height: 48px;"
          >
            <p
              class="MuiTypography-root MuiTypography-body2"
            >
              Type
               
              
            </p>
          </div>
        </div>
        <div
          class="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style="flex: 0 1 110px;"
        >
          <div
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiVirtualizedTable-tableCell-4 MuiVirtualizedTable-flexContainer-1"
            style="height: 48px;"
          >
            <p
              class="MuiTypography-root MuiTypography-body2"
            >
              Size
               
              
            </p>
          </div>
        </div>
        <div
          class="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style="flex: 0 1 180px;"
        >
          <div
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiVirtualizedTable-tableCell-4 MuiVirtualizedTable-flexContainer-1"
            style="height: 48px;"
          >
            <p
              class="MuiTypography-root MuiTypography-body2"
            >
              Assigned To
               
              
            </p>
          </div>
        </div>
        <div
          class="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style="flex: 0 1 210px;"
        >
          <div
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiVirtualizedTable-tableCell-4 MuiVirtualizedTable-flexContainer-1"
            style="height: 48px;"
          >
            <p
              class="MuiTypography-root MuiTypography-body2"
            >
              Location
               
              
            </p>
          </div>
        </div>
        <div
          class="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style="flex: 0 1 120px;"
        >
          <div
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiVirtualizedTable-tableCell-4 MuiVirtualizedTable-flexContainer-1"
            style="height: 48px;"
          >
            <p
              class="MuiTypography-root MuiTypography-body2"
            >
              Condition
               
              
            </p>
          </div>
        </div>
        <div
          class="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style="flex: 0 1 110px;"
        >
          <div
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiVirtualizedTable-tableCell-4 MuiVirtualizedTable-flexContainer-1"
            style="height: 48px;"
          >
            <p
              class="MuiTypography-root MuiTypography-body2"
            >
              Quality
               
              
            </p>
          </div>
        </div>
        <div
          class="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style="flex: 0 1 100px;"
        >
          <div
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiVirtualizedTable-tableCell-4 MuiVirtualizedTable-flexContainer-1"
            style="height: 48px;"
          >
            <p
              class="MuiTypography-root MuiTypography-body2"
            >
              Gifted
               
              
            </p>
          </div>
        </div>
        <div
          class="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style="flex: 0 1 200px;"
        >
          <div
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiVirtualizedTable-tableCell-4 MuiVirtualizedTable-flexContainer-1 MuiVirtualizedTable-noClick-5"
            style="height: 48px;"
          >
            <p
              class="MuiTypography-root MuiTypography-body2"
            >
              History
               
              
            </p>
          </div>
        </div>
      </div>
      <div
        aria-label="grid"
        aria-readonly="true"
        class="ReactVirtualized__Grid ReactVirtualized__Table__Grid"
        role="rowgroup"
        style="box-sizing: border-box; direction: ltr; height: -48px; position: relative; width: 0px; will-change: transform; overflow-x: hidden; overflow-y: auto;"
        tabindex="0"
      />
    </div>
  </div>
  <div
    class="resize-triggers"
  >
    <div
      class="expand-trigger"
    >
      <div
        style="width: 1px; height: 1px;"
      />
    </div>
    <div
      class="contract-trigger"
    />
  </div>
</div>
`;
